packages/std/src/imports.rs

```diff
@@ -129,16 +133,89 @@ impl Storage for ExternalStorage {
         end: Option<&[u8]>,
         order: Order,
     ) -> Box<dyn Iterator<Item = Record>> {
-        // There is lots of gotchas on turning options into regions for FFI, thus this design
-        // See: https://github.com/CosmWasm/cosmwasm/pull/509
-        let start_region = start.map(build_region);
-        let end_region = end.map(build_region);
-        let start_region_addr = get_optional_region_address(&start_region.as_ref());
-        let end_region_addr = get_optional_region_address(&end_region.as_ref());
-        let iterator_id = unsafe { db_scan(start_region_addr, end_region_addr, order as i32) };
+        let iterator_id = create_iter(start, end, order);
         let iter = ExternalIterator { iterator_id };
         Box::new(iter)
     }
+
+    #[cfg(all(feature = "cosmwasm_1_4", feature = "iterator"))]
+    fn range_keys<'a>(
+        &'a self,
+        start: Option<&[u8]>,
+        end: Option<&[u8]>,
+        order: Order,
+    ) -> Box<dyn Iterator<Item = Vec<u8>> + 'a> {
+        let iterator_id = create_iter(start, end, order);
+        let iter = ExternalPartialIterator {
+            iterator_id,
+            partial_type: PartialType::Keys,
+        };
+        Box::new(iter)
+    }
+
+    #[cfg(all(feature = "cosmwasm_1_4", feature = "iterator"))]
+    fn range_values<'a>(
+        &'a self,
+        start: Option<&[u8]>,
+        end: Option<&[u8]>,
+        order: Order,
+    ) -> Box<dyn Iterator<Item = Vec<u8>> + 'a> {
+        let iterator_id = create_iter(start, end, order);
+        let iter = ExternalPartialIterator {
+            iterator_id,
+            partial_type: PartialType::Values,
+        };
+        Box::new(iter)
+    }
+}
+
+#[cfg(feature = "iterator")]
+fn create_iter(start: Option<&[u8]>, end: Option<&[u8]>, order: Order) -> u32 {
+    // There is lots of gotchas on turning options into regions for FFI, thus this design
+    // See: https://github.com/CosmWasm/cosmwasm/pull/509
+    let start_region = start.map(build_region);
+    let end_region = end.map(build_region);
+    let start_region_addr = get_optional_region_address(&start_region.as_ref());
+    let end_region_addr = get_optional_region_address(&end_region.as_ref());
+    unsafe { db_scan(start_region_addr, end_region_addr, order as i32) }
+}
+
+#[cfg(all(feature = "cosmwasm_1_4", feature = "iterator"))]
+enum PartialType {
+    Keys,
+    Values,
+}
+
+/// ExternalPartialIterator makes a call out to `next_key` or `next_value`
+/// depending on its `partial_type`.
+/// Compared to `ExternalIterator`, it allows iterating only over the keys or
+/// values instead of both.
+#[cfg(all(feature = "cosmwasm_1_4", feature = "iterator"))]
+struct ExternalPartialIterator {
+    iterator_id: u32,
+    partial_type: PartialType,
+}
```

Comment:
I was thinking about implementing Iterator for the enum directly:

```rust
enum ExternalPartialIterator {
    Keys { iterator_id: u32 },
    Values { iterator_id: u32 },
}
```

However, I don't see any strong benefit and it makes it harder to add fields, so feel free to ignore if not useful.